### ISSUE TITLE
Expose docker on TCP port cause crash with Docker Desktop

### ISSUE THREAD

**Issue by masoudr** [docker/for-win#6895]
- [ ] I have tried with the latest version of my channel (Stable or Edge)
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Expected behavior
The service should set up without any problem.
### Actual behavior
Here is the error I get
```
Docker.Core.Backend.BackendException:
error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.40/containers/json: open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running.

   at Docker.Core.Pipe.NamedPipeClient.<TrySendAsync>d__5.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Docker.Core.Pipe.NamedPipeClient.Send(String action, Object[] parameters)
   at Docker.Actions.DoRestartVM(Settings previousSettings)
   at Docker.Actions.<>c__DisplayClass34_0.<RestartVM>b__0()
   at Docker.ApiServices.TaskQueuing.TaskQueue.<>c__DisplayClass18_0.<.ctor>b__1()

  Docker.Core.DockerException:
  error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.40/containers/json: open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running.
  
     at Docker.Backend.DockerDaemonChecker.Check(Func`1 isDaemonProcessStillRunning)
     at Docker.Backend.ContainerEngine.Windows.DoStart(Settings settings, String daemonOptions)
     at Docker.Backend.ContainerEngine.Windows.Restart(Settings settings, String daemonOptions, Credential credential)
     at Docker.Backend.BackendNamedPipeServer.<Run>b__8_3(Object[] args)
     at Docker.Core.Pipe.NamedPipeServer.<>c__DisplayClass9_0.<Register>b__0(Object[] parameters)
     at Docker.Core.Pipe.NamedPipeServer.RunAction(String action, Object[] parameters)
  
```
### Information

  - Windows Version: Win10 1909 X64
  - Docker Desktop Version: 2.2.0.3
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: No

I'm using Windows container on Docker Desktop. I'm trying to expose docker daemon on TCP port without TLS but when I set the option in Docker settings I get the above exception.

### Steps to reproduce the behavior

Just set the checkbox in setting `Expose daemon on TCP://localhost:2375 without TLS`.
Or setting it in daemon.json:
```
{
    "hosts": ["tcp://0.0.0.0:2375"]
}
```

**Comment by StefanScherer** [comment]
@masoudr The `hosts` setting in the `daemon.json` won't work and leads to an error starting Docker Desktop. This disables the default Unix socket that is needed for the Docker Desktop proxy we use to share it to Windows and.

Please remove this from the Docker Engine settings and try to restart Docker Desktop.

My setup is `Use the WSL 2 based engine` enabled and `Expose daemon on tcp://localhost:2375 without TLS` enabled. We investigate further what might cause the problem.

**Comment by masoudr** [comment]
@StefanScherer Thanks for the info. I tried to just set the `Expose daemon on tcp://localhost:2375 without TLS` option but I got the same behavior. Also, the `Use the WSL 2 based engine` option is grayed out for me. Do you have any idea how to set it?

**Comment by StefanScherer** [comment]
Oh yes, the WSL 2 based engine option is grayed out, because you are on Win10 1909 X64, I haven't noticed that earlier. WSL 2 is only available with Win10 2004.

**Comment by masoudr** [comment]
@StefanScherer Thanks for the info. I migrated to WSL2 and the problem solved.

**Comment by stephen-turner** [comment]
FYI The underlying bug is fixed in the latest Edge release, 2.3.2.0.

**Comment by masoudr** [comment]
@stephen-turner Thanks for the update. I wanted to use this feature in the Windows container as it doesn't support the special docker DNS name.

**Comment by PavelSosin-320** [comment]
In 2.3.2.1 this issue is not solved yet. But numerous new issues popped up with the latest Windows upgrade. I have build 20.161 and only WSL distros added after the upgrade are running. The bright side - with correct resolv.conf reflecting my router DNS setup.
Numerous changes I detected in the new WSL version, for example how it starts up. I also found that my Windows environment contains the wrong path to Program Files/Docker/Docker/resources/bin inherited by Docker Desktop distro.
